### PR TITLE
[IE CLDNN] Added batch_to_space operation

### PR DESCRIPTION
**Pull request by Lyamin-Roman**

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/api/batch_to_space.hpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+namespace cldnn {
+/// @addtogroup cpp_api C++ API
+/// @{
+/// @addtogroup cpp_topology Network Topology
+/// @{
+/// @addtogroup cpp_primitives Primitives
+/// @{
+
+/// @brief
+/// @details
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/api/batch_to_space.hpp:30]: You should write brief & details information here.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    const uint feature = get_global_id(1);
+
+#ifdef OUTPUT_LAYOUT_BFYX
+    const uint w = 0;
+    const uint z = 0;
+    const uint y = get_global_id(2) / OUTPUT_SIZE_X;
+    const uint x = get_global_id(2) % OUTPUT_SIZE_X;
+#elif OUTPUT_LAYOUT_BFZYX
+    const uint w = 0;
+	const uint yx = get_global_id(2) % (OUTPUT_SIZE_X * OUTPUT_SIZE_Y);
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl:30]: Bad align

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    const uint offset_y = (y + CROPS_BEGIN_Y) % BLOCK_SHAPE_Y;
+
+    const uint input_x = (x + CROPS_BEGIN_X) / BLOCK_SHAPE_X;
+    const uint offset_x = (x + CROPS_BEGIN_X) % BLOCK_SHAPE_X;
+
+    const uint offset_batch = (offset_feature * BLOCK_SHAPE_W * BLOCK_SHAPE_Z * BLOCK_SHAPE_Y * BLOCK_SHAPE_X +
+                            offset_w * BLOCK_SHAPE_Z * BLOCK_SHAPE_Y * BLOCK_SHAPE_X +
+                            offset_z * BLOCK_SHAPE_Y * BLOCK_SHAPE_X +
+                            offset_y * BLOCK_SHAPE_X +
+                            offset_x) * OUTPUT_BATCH_NUM;
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl:62]: You should align all this lines with previous (offset feature).

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    const uint input_batch = batch + offset_batch;
+
+    const uint input_index = INPUT0_OFFSET +
+            input_batch * INPUT0_BATCH_PITCH +
+            input_feature * INPUT0_FEATURE_PITCH +
+            input_w * INPUT0_W_PITCH +
+            input_z * INPUT0_Z_PITCH +
+            input_y * INPUT0_Y_PITCH +
+            input_x;
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl:71]: And same here.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    const uint output_index = OUTPUT_OFFSET +
+            batch * OUTPUT_BATCH_PITCH +
+            feature * OUTPUT_FEATURE_PITCH +
+            w * OUTPUT_W_PITCH +
+            z * OUTPUT_Z_PITCH +
+            y * OUTPUT_Y_PITCH +
+            x;
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl:79]: Same here.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/batch_to_space/batch_to_space_kernel_selector.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/*
+// Copyright (c) 2019 Intel Corporation
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/batch_to_space/batch_to_space_kernel_selector.h:2]: Please change year in the license info.

File: inference-engine/thirdparty/clDNN/src/batch_to_space.cpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    for(size_t i = 1; i < dims_num; ++i)
+        if (crops_begin[i] + crops_end[i] >= block_shape[i] * input_layout.size.sizes(input_format)[i])
+            CLDNN_ERROR_MESSAGE(node.id(),
+                "Output dimensions must be positive");
+
+    const auto output_batch = input_layout.size.batch[0] / block_sizes_multiplied;
+
+    std::vector<int32_t> output_shape;
+    output_shape.reserve(dims_num);
+    for (size_t i = 1; i < dims_num; ++i){
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/src/batch_to_space.cpp:89]: Space before brace.

File: inference-engine/thirdparty/clDNN/src/batch_to_space.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+#include "data_inst.h"
+#include <string>
+#include <vector>
+
+namespace cldnn {
+primitive_type_id cldnn::batch_to_space::type_id() {
+    static primitive_type_base<batch_to_space> instance;
+    return &instance;
+}
+
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/src/batch_to_space.cpp:31]: Empty line

File: inference-engine/thirdparty/clDNN/src/batch_to_space.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    std::vector<int32_t> output_shape;
+    output_shape.reserve(dims_num);
+    for (size_t i = 1; i < dims_num; ++i) {
+        output_shape.emplace_back(input_layout.size.sizes(input_format)[i] * block_shape[i] - crops_begin[i] - crops_end[i]);
+    }
+
+    switch(input_format) {
+        case format::bfzyx:
+        return layout {
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/src/batch_to_space.cpp:95]: Please use tab at the next line after case statement: 
```
        case format::bfzyx:
            return layout {
```

File: inference-engine/thirdparty/clDNN/src/gpu/batch_to_space_gpu.cpp [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        return batch_to_space;
+    }
+};
+
+namespace detail {
+
+attach_batch_to_space_gpu::attach_batch_to_space_gpu() {
+    auto val_fw = batch_to_space_gpu::create;
+    implementation_map<batch_to_space>::add(std::make_tuple(engine_types::ocl, data_types::f32, format::bfyx),
+                                            val_fw);
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/src/gpu/batch_to_space_gpu.cpp:71]: I think single line will be more readable:
```
implementation_map<batch_to_space>::add(std::make_tuple(engine_types::ocl, data_types::f32, format::bfyx), val_fw);
implementation_map<batch_to_space>::add(std::make_tuple(engine_types::ocl, data_types::f16, format::bfyx), val_fw);
```

File: inference-engine/thirdparty/clDNN/tests/test_cases/batch_to_space_gpu_test.cpp [hunk 11]
```diff
@@ -0,0 +1,651 @@
...
+    };
+
+    ASSERT_EQ(output_ptr.size(), expected_results.size());
+
+    for (size_t i = 0; i < expected_results.size(); ++i) {
+        EXPECT_EQ(expected_results[i], output_ptr[i]);
+    }
+}
+
+
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/tests/test_cases/batch_to_space_gpu_test.cpp:463]: nit: empty line

File: inference-engine/thirdparty/clDNN/tests/test_cases/batch_to_space_gpu_test.cpp [hunk 12]
```diff
@@ -0,0 +1,651 @@
...
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/tests/test_cases/batch_to_space_gpu_test.cpp:525]: nit: empty line

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+    const uint output_index = OUTPUT_OFFSET +
+                              batch * OUTPUT_BATCH_PITCH +
+                              feature * OUTPUT_FEATURE_PITCH +
+                              w * OUTPUT_W_PITCH +
+                              z * OUTPUT_Z_PITCH +
+                              y * OUTPUT_Y_PITCH +
+                              x;
+
+    output[output_index] = ACTIVATION(input[input_index], ACTIVATION_PARAMS);
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/batch_to_space_ref.cl:81]: `ACTIVATION `macro do nothing until you add your primitive into "White list" at `void prepare_primitive_fusing::fuse_activations(program_impl &p)` function in `prepare_primitive_fusing.cpp`

File: inference-engine/thirdparty/clDNN/api/batch_to_space.hpp [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+struct batch_to_space : public primitive_base<batch_to_space> {
+    CLDNN_DECLARE_PRIMITIVE(batch_to_space)
+
+    /// @brief Constructs batch_to_space primitive.
+    /// @param id This primitive id.
+    /// @param input Input dictionary primitive id.
+    /// @param block_shape Array of block sizes.
+    /// @param crops_begin Amount to crop from the beginning along each axis of data input.
+    /// @param crops_end Amount to crop from the ending along each axis of data input.
+
```
Comment (echuraev) [inline inference-engine/thirdparty/clDNN/api/batch_to_space.hpp:65]: Param in doxygen comments are not the same with real names in function. Also, please remove empty line.

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp [hunk 15]
```diff
@@ -82,7 +82,9 @@ InferenceEngine::ICNNNetwork::Ptr clDNNEngine::CloneNetwork(const InferenceEngin
             }
 
             return std::dynamic_pointer_cast<const ::ngraph::opset2::Gelu>(node) ||
-                   std::dynamic_pointer_cast<const ::ngraph::opset3::ShuffleChannels>(node);
+                   std::dynamic_pointer_cast<const ::ngraph::opset3::ShuffleChannels>(node) ||
+                   std::dynamic_pointer_cast<const ::ngraph::opset2::BatchToSpace>(node) ||
+                   std::dynamic_pointer_cast<const ::ngraph::opset2::SpaceToBatch>(node);
```
Comment (vladimir-paramuzov) [inline inference-engine/src/cldnn_engine/cldnn_engine.cpp:87]: SpaceToBatch is not implemented yet. please delete this line

**Pull request merged**